### DISCUSSION TITLE
[BE] soft delete하는데 꼭 CouponDesign, CouponPolicy의 복사본이 필요할까?(by 레오)

### DISCUSSION THREAD

**Discussion by yenawee** [woowacourse-teams/2023-stamp-crush#238] [General]
불필요하게 코드가 복잡해지는 느낌이 들었습니다.

**Comment by gitchannn** [comment]
**하디**: soft delete를 고려했던 것이 아니라, 데이터베이스 상에서의 테이블에서 사이클이 돌아서 이렇게 만들었던 것이다. 하지만, 객체 상에서는 사이클이 돌지 않는다. 
**레오**: 데이터베이스 상에서 보면 fk만 걸려있어도 양방향이다. 데이터베이스 상에서는 단방향이라는게 없다. 데이터베이스에서 사이클을 고려할 것이 아닌 것 같다.

**Comment by gitchannn** [comment]
## 깃짱: 레벨4이던가 4차 데모데이이던가.. 데이터베이스 마이그레이션을 해보는 게 좋을 것 같다.

**Comment by gitchannn** [comment]
# migration을 해볼 수 있으니 너무 안좋게 생각하지 맙시다!

**Reply by gitchannn** [reply]
## 복사본을 없앤다.

### 없앨 테이블
- coupon_design
- coupon_policy
- coupon_stamp_coordinate

### `Coupon` 테이블 자체에 추가할 fk
- cafe_coupon_design_id
- cafe_policy_id

### 추가 논의사항
- 현재 coupon_design, coupon_policy에서 String 값만 복사해오고, 아이디값을 전혀 가져오지 않는 상태. 지금까지 가지고 있는 쿠폰에서 어떻게 해결할 것인가
- 현재 카페 사장이 쿠폰 디자인이나 정책을 변경하게 되면, coupon_design, coupon_policy를 soft delete 해버리는데, 그래서 억지로 jpql을 작성해서 execute하지 않으면 가져오지 않을 것임. 
    - outdated 컬럼 추가
    - 정책이나 디자인이 변경되면, delete 하는 것이 아니라 **outdated 필드를 update하는 방향으로 수정**
    - 이제까지 만들었던 모든 카페 디자인과 정책에 대해서 deleted=false로 변경
